### ISSUE TITLE
(Feature-ish request) Hash texture names at compile time?

### ISSUE THREAD

**Issue by TorutheRedFox** [xan1242/NFS-XtendedInput#23]
This would speed things up a bit I think (along with the implementation of this into the TPK tool so you don't have to do it manually in the INI)

dunno how to implement the algorithm into a macro though, but thankfully C++ 11 has our backs and allows us to just write constant expressions which have the same end result as a macro while having more flexibility as you no longer need hacks to do recursion (which has to be fixed length too)

here's a quick matching implementation of it (you can also make a non-constexpr copy of it for runtime computation of hashes, could be useful in giving the different controllers string names in the INI without needing to rewrite much maybe?)

```C++
constexpr unsigned int BIN_HASH(const char* string)
{
    char* _string = ((char*)string);
    int result = -1;

    while (*_string != 0) {
        result = result * 0x21 + (unsigned int)(*_string);
        _string = _string + 1;
    }

    return (unsigned int)result;
}
```

**Comment by TorutheRedFox** [comment]
i.e. rather than
```C++
#define PS4_DPAD_UP_HASH 0x52FB3FD2
```
you'd have
```C++
#define PS4_DPAD_UP_HASH BIN_HASH("PS4_DPAD_UP")
```
and it'd expand to 0x52FB3FD2 at compile time

**Comment by xan1242** [comment]
Hmmm that would be useful to do during compile time. Good idea!

I wasn't very familiar with constexpr, to be perfectly honest with you, but I could implement this.

I currently use the ripped bStringHash function but this appears to be the exact same thing anyways, so if you want to, you may submit a PR for this.

Or I'll do it myself, your choice.

**Comment by TorutheRedFox** [comment]
it's a version of bStringHash I decompiled with Ghidra then manually cleaned up to remove redundancy

**Comment by TorutheRedFox** [comment]
FYI I'm working on making the hash macros use this instead
I'll open a PR once I'm done

**Comment by TorutheRedFox** [comment]
Opened pull request at https://github.com/xan1242/NFS-XtendedInput/pull/24

**Comment by xan1242** [comment]
With [0995a41](https://github.com/xan1242/NFS-XtendedInput/commit/0995a41b893f74638839f500c58ae3165306eef3) I'm considering this implemented, closing.

**Comment by xan1242** [comment]
Nuh-uh, after further snooping, constexpr doesn't do its job. Reverting all changes.

**Comment by xan1242** [comment]
![image](https://user-images.githubusercontent.com/8014093/167252722-b29cb59d-b882-4399-aadc-34314c15ec0d.png)

Proof here: IDA shows this in GetAppropriateButtonTex function. This is not right and it should NOT happen.

**Comment by xan1242** [comment]
![image](https://user-images.githubusercontent.com/8014093/167252838-599e9eac-8ca1-4497-adb0-f1e1a61f716f.png)

It should look a little something like this

**Comment by TorutheRedFox** [comment]
hmmm

that's odd

**Comment by TorutheRedFox** [comment]
I'll look into what's causing this and if it can be fixed

otherwise, time to make that macro then :D

**Comment by xan1242** [comment]
I've tested with Carbon but that doesn't have a barring on this, this is the compiler's doing.

**Comment by xan1242** [comment]
There isn't a way to make this a "macro" per se, from what I've read. Maybe I could be wrong, but constexpr is the way to go.

**Comment by xan1242** [comment]
The solution is to convert the hash macros into enums instead. That way it'll actually get processed by the compiler.

**Comment by TorutheRedFox** [comment]
I think I fixed it with a dirty hack?

**Comment by TorutheRedFox** [comment]
```C++
// compile time hasher thanks to C++11
// I tried making a macro version but failed xdddd
// P.S. it causes IntelliSense to freak out in some cases, but it still compiles
static constexpr int __HASH(char* string)
{
	if (string == NULL) // sanity check
		return 0;

	char* _string = ((char*)string);
	int result = -1;

	while (*_string != 0) { // loop through each char until string terminator is reached
		result = result * 0x21 + (unsigned int)(*_string);
		_string = _string + 1;
	}

	return result;
}

template <int V> static constexpr int _HASH = V;

#define HASH(str) _HASH<__HASH(str)>
```
it seems to work, as I don't see any instances of the source strings in the output binary now

**Comment by TorutheRedFox** [comment]
tested ingame and yeah it works

**Comment by xan1242** [comment]
Result is OK in this case as well. Either way would work (enums or templates).

**Comment by xan1242** [comment]
implemented in [666286e](https://github.com/xan1242/NFS-XtendedInput/commit/666286e0392fa5d459d4c875a8459f83e08e8ec3), closing
